astropy/modeling/astro_sherpa.py

```diff
@@ -0,0 +1,433 @@
+            raise TypeError("statistic must be sherpa Stat class or isinstance")
+
+        if inspect.isclass(estmethod) and not issubclass(estmethod, EstMethod):
+            raise TypeError("estmethod must be sherpa EstMethod class or isinstance")
+        elif inspect.isfunction(estmethod) and not issubclass(estmethod.__class__, EstMethod):
+            raise TypeError("estmethod must be sherpa EstMethod class or isinstance")
+
+        super(SherpaFitter, self).__init__(optimizer=optimizer, statistic=statistic)
+        if inspect.isclass(estmethod):
+            self._est_method = estmethod()
+        else:
+            self._est_method = estmethod
+
+        self.fit_info = {}
+        self._fitter = None  # a handle for sherpa fit function
+        self._fitmodel = None  # a handle for sherpa fit model
+        self._data = None  # a handle for sherpa dataset
+
+    def __call__(self, models, x, y, z=None, xerr=None, yerr=None, zerr=None, **kwargs):
+        """
+        Fit the astropy model with a the sherpa fit routines.
+
+        Parameters
+        ----------
+        model : `~astropy.modeling.FittableModel`
+            model to fit to x, y, z
+        x : array
+            input coordinates
+        y : array
+            input coordinates
+        z : array (optional)
+            input coordinates
+        xerr : array (optional)
+            an array of errors in x
+        yerr : array (optional)
+            an array of errors in y
+        zerr : array (optional)
+            an array of errors in z
+        **kwargs:
+            keyword arguments will be passed on to sherpa fit routine
+
+        Returns
+        -------
+        model_copy : `~astropy.modeling.FittableModel` or a list of models.
+            a copy of the input model with parameters set by the fitter
+        """
+
+        model_dict = OrderedDict()
+        tie_list = []
+
+        if isinstance(models, (FittableModel, _CompoundModel)):
+            models = [models]
+
+        self._data = make_datasets(models[0].n_inputs, x, y, z, xerr, yerr, zerr)
+
+        if isinstance(self._data, DataSimulFit):
+            _ndata = len(self._data.datasets)
+            if len(models) == 1:
+                self._fitmodel = _make_simfit_model([models[0].copy() for _ in xrange(_ndata)], tie_list, model_dict)
+                # Copy the model so each data set has the same model!
+            elif len(models) == _ndata:
+                self._fitmodel = _make_simfit_model(models, tie_list, model_dict)
+            else:
+                raise Exception("Don't know how to handle multiple models unless there is one foreach dataset")
+        else:
+            if len(models) > 1:
+                self._data = DataSimulFit("wrapped_data", [self._data for _ in xrange(len(models))])
+                self._fitmodel = _make_simfit_model(models, tie_list, model_dict)
+            else:
+                self._fitmodel = _astropy_to_sherpa_model(models[0])
```

Comment:
This is where having the wrapper classes I was talking about should help you: this block of code is building the model, which depends on what kind of dataset you are working with: rather than having the `isinstance` calls, you could just call a method in the wrapper dataset classes.